airflow/jobs/scheduler_job_runner.py

```diff
@@ -422,7 +423,15 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session) -
             task_instance_str = "\n".join(f"\t{x!r}" for x in task_instances_to_examine)
             self.log.info("%s tasks up for execution:\n%s", len(task_instances_to_examine), task_instance_str)
 
+            tis_we_have_room_for = self._slots_free_for_tis(task_instances_to_examine)
```

Comment:
Ok so, further down, we apply more constraints like, pools, dag concurrency, task concurrency.  So I can imagine a circumstance, though perhaps rare, where the scheduler cannot schedule anything because there are let's say 10 slots avail in the executor, but those 10 tasks can't run because of the checks that are applied further down.  And then it keeps processing the same TIs in each loop?  Perhaps we should not allocate a slot to a TI until it passes all conditions?

Reply:
Hey @dstandish, good call out! The scheduler loop already has provisions for this and I think the best fix is to keep this filtering where it is with all the others in this for loop, but to add the tasks to the starved_tasks list if they can't be scheduled because the executor doesn't have room. That way they'll be excluded from the query next loop (so we won't keep processing the same TIs), just as we do for pools and and dag concurrency etc.

What do you think?

Reply:
i think it's fine to keep the other filters where they are.  it seemed like the thing to do is just don't decrement the available slots til you add the TI to executable TIs.

so first you would create this dict

```
slots_available = {x.name: x.slots_available for x in self.job.executors}
```

e.g. immediately after this line
```python
executable_tis: list[TI] = []
```

then immediately before or after this line....
```python
executable_tis.append(task_instance)
```

you would do something like 
```python
slots_available[task_instance.executor] -= 1
```

of course you'd need to handle normalizing the name and ensuring that it doesn't negative. i don't know how "starved_tasks" works and what the implications of that is and e.g. whether you should add it in the "not enough slots" case.

anyway, that's just how it looked to me. you have been engaging with this much more deeply than I have though so you may have better solution.

Reply:
Yupp, I see where you're saying, it is slightly more optimal in decreasing the odds that you'll have to do another loop to fetch more tasks, but I think we're talking about very small margins here. I'll play around with the code changes and if it doesn't add too much complexity then it might be worth doing to get those marginal gains :+1: